titan-compiler/checker.lua

```diff
@@ -93,11 +100,67 @@ local function coerce_numeric_exp_to_float(exp)
     end
 end
 
+-- Does this statement always call "return"?
+--
+-- In the future I would like to get rid of the function, and make it a part of
+-- live-variable analysis. (A possibly-uninitialized "return variable" signifies
+-- a missing return statement.)
+local function stat_always_returns(stat)
+    local tag = stat._tag
+    if     tag  == ast.Stat.Decl then
+        return false
+    elseif tag == ast.Stat.Block then
+        for _, inner_stat in ipairs(stat.stats) do
+            if stat_always_returns(inner_stat) then
+                return true
+            end
+        end
+        return false
+    elseif tag == ast.Stat.While then
+        return false
+    elseif tag == ast.Stat.Repeat then
+        return false
+    elseif tag == ast.Stat.For then
+        return false
+    elseif tag == ast.Stat.Assign then
+        return false
+    elseif tag == ast.Stat.Call  then
+        return false
+    elseif tag == ast.Stat.Return then
+        return true
+    elseif tag == ast.Stat.If then
+        for _, thn in ipairs(stat.thens) do
+            if not stat_always_returns(thn.block) then
+                return false
+            end
+        end
+        if not stat.elsestat or not stat_always_returns(stat.elsestat) then
+            return false
+        end
+        return true
+    else
+        error("impossible")
+    end
+end
+
 --
 -- check
 --
 
-check_program = function(prog, errors)
+typecheck = function(prog)
+    local ok, err = xpcall(check_program, debug.traceback, prog)
+    if ok then
+        return prog, {} -- TODO:  no {}
+    else
+        if getmetatable(err) == TypeError then
+            return false, { err.msg } -- TODO  no {}
+        else
+            error(err)
```

Comment:
If we call error here it will print two stack traces. We probably should do:
```
io.stderr:write(err, '\n')
os.exit(1)
```

Reply:
I'm not fond of using os.exit in an internal module like this one. And if we don't use os.exit I think the duplicate stack trace is unavoidable.

Reply:
Indeed.